### DISCUSSION TITLE
Edge detection & platformer logic coding

### DISCUSSION THREAD

**Discussion by Ace4teaM** [AdamsLair/duality#894] [Questions & Help]
Hello everyone, I am new to the Duality community and I have developed a remake of the game Front Mission gun hazard from the Super-Nintendo. For the moment I am starting, I discovered Duality 2 days ago and I totally hang on to the foundation of the software compared to the masses that are Unity and Unreal. Fast learning curve and the ability to develop C# components, love it!

Anyway… today I'm having fun with the character's movements and the physics system. I have some ideas for further development and maybe a more "modern" component than map tiles for composing platforms.

I have a few questions for the sequel, I would like a 2D platformer with good Mario style gameplay and not just pool ball style collisions.

Character states to be determined:
•	On air
•	On the ground
•	On the wall
•	OnSlide
•	OnDown

![b](https://user-images.githubusercontent.com/2634510/151656668-72846c79-758f-4d2b-a343-ec11746de579.jpg)

To code all these states I want to rely on the physics system but not that, for example I don't want the character to slide down the stairs, I want him to go down the steps with an animation.
For this I want to detect which RigidBody my character is in collision with. Currently I'm using the Scene.Physics.QueryRect method but maybe you have a better idea.

Simple technique: if character's body + 1, then collision:
#
```
var list = new List<RigidBody>();
this.Scene.Physics.QueryRect(trans.Pos.Xy, new Vector2(body.BoundRadius, body.BoundRadius), list);
this.ground = list.Count > 1;

if (this.ground)
{
	// on ground
}
else
{
	// On air
}

```
1. How to know the edge and the angle of the segment on which the character is?

![a](https://user-images.githubusercontent.com/2634510/151656679-efe1d03a-3d44-4ab2-98be-5e422f923b51.jpg)

Sorry for my English which is not perfect. I use google translation from French do not hesitate to ask me questions if necessary!

Thank you all

The original game !
![164334-front-mission-gun-hazard-snes-screenshot-trying-to-defend](https://user-images.githubusercontent.com/2634510/151656798-dc61dba5-d382-4676-bf33-a2f088c412b1.png)

**Comment by SirePi** [comment]
Hi! sorry for the delay :)
I don't know if it's what you need, but maybe you could have your character's GameObject have multiple sensor shapes scattered around, and based on which and how many are triggered, you can detect the situation? i.e. no sensor triggers = in air; bottom sensor triggers = on ground; left/right = on wall...
Don't know how it would work for "going up/down".. maybe having one sensor on each foot?
